R/simulateStochasticLogistic.R

```diff
@@ -0,0 +1,151 @@
+#'\linkS4class{SummarizedExperiment} object.
+#'
+#' @param n.species Integer: number of species
+#' @param t.start Numeric scalar indicating the initial time of the simulation.
+#' (default: \code{t.start = 0})
+#' @param t.end Numeric scalar indicating the final time of the dimulation
+#' (default: \code{t.end = 2000})
+#' @param t.step Numeric scalar indicating the interval between simulation steps
+#' (default: \code{t.step = 0.01})
+#' @param t.store Integer scalar indicating the number of evenly distributed
+#' time points to keep
+#' (default: \code{t.store = 1000})
+#' @param partial Logical: whether to export only the stored time points
+#' (default: \code{partial = TRUE})
+#' @param stochastic Logical: whether the logistic model should be stochastic
+#' (controlled by multiplying the growth rate by a random number)
+#' (default: \code{stochastic = TRUE})
+#'
+#' @docType methods
+#' @examples
+#' ## ATTENTION: Don't set a large value to t.step, otherwise the computer won't
+#' give a correct solution to the logistic ODE(ordinary differential equation).
+#' Keeping t.step under 0.05 or 0.01 is a good practice.
+#'
+#' while (!exists("ExampleLogistic"))
+#' ExampleLogistic <- simulateStochasticLogistic(n.species = 5)
+#' # plot the calculated points
+#' matplot(t(ExampleLogistic@assays@data@listData[["counts"]]), type = "l")
+#'
+#' # calculation by setting initial parameters explicitly
+#' ExampleLogistic <- simulateStochasticLogistic(n.species = 2, b <- c(0.2, 0.1), k <- c(1000, 2000), dr <- c(0.001, 0.0015), x <- c(3, 0.1), t.start = 0, t.end = 1500, t.step = 0.01, t.store = 1200, partial = F, stochastic = F)
+#'
+#' @return \code{simulateStochasticLogistic} returns a
+#' \linkS4class{SummarizedExperiment} class object containing matrix with
+#' species abundance as rows and time points as columns
+#'
+#' @export
+
+setGeneric("simulateStochasticLogistic",signature = "n.species",
+        function(n.species,
+                t.start = 0,
+                t.end = 2000,
+                t.step = 0.01,
+                t.store = 1000,
+                partial = TRUE,
+                stochastic = TRUE)
+        standardGeneric("simulateStochasticLogistic"))
+
+setMethod("simulateStochasticLogistic", signature = c(n.species="numeric"),
+        function(n.species,
+                t.start = 0,
+                t.end = 2000,
+                t.step = 0.01,
+                t.store = 1000,
+                partial = TRUE,
+                stochastic = TRUE){
+        #  growth rates
+        b <-  runif(n = n.species, min = 0.1, max = 0.2)
+        # carrying capacities
+        k <- runif(n = n.species, min = 1000, max = 2000)
+        # death rates
+        dr <- runif(n = n.species, min = 0.0005, max = 0.0025)
+        # initial abundances
+        x <- runif(n = n.species, min = 0.1, max = 10)
+
+        # define the logistic model
+        logistic.model <- function (t, state, params) {
```

Comment:
Nested functions. A good name would be `.sim_time_logistic_model` Avoid `.` in the function name, because the can be condussed with S3 dispatch requirements